### DISCUSSION TITLE
Roadmap for Node-RED dashboard running on MCU

### DISCUSSION THREAD

**Discussion by rei-vilo** [phoddie/node-red-mcu#40] [General]
One of the greatest and possibly most popular features of Node-RED is the dashboard. 

The video from the Node-RED Con 2022 shows at [2:19:41](https://youtu.be/Inf37X0_xxo?t=8381) a Node-RED dashboard running on an MCU.

Is there a roadmap (or OTA) for running Node-RED dashboard on an MCU?

Thank you!

**Comment by ralphwetzel** [comment]
The dashboard already runs on an MCU - as @phoddie has demonstrated in the mentioned session. What's missing - currently - is the support by the plugin. You yet might have noticed the first traces that it is in the making:

<img width="283" alt="image" src="https://user-images.githubusercontent.com/16342003/201170662-a106b479-aef2-43b3-81d4-61cc008204a2.png">

There's some work still to be done; thus I need to ask you for some (additional) patience...

**Comment by rei-vilo** [comment]
Thank you for your prompt answer and good luck with the development!

**Reply by phoddie** [reply]
Regarding what hardware you need to run the dashboard, there are some options. A great starting point is the simulator: no extra hardware required at all! The dashboard requires a fair amount of memory, so an ESP8266 is impractical. An ESP32 is good and a Raspberry Pi Pico will work too (though it has less RAM, so you will hit a limit sooner). As you might imagine, we use [Moddable Two](https://www.moddable.com/product.php) for our development. So, if you are looking to keep things simple, that's the best starting point. The Moddable SDK has extensive lists of supported [ESP32](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/devices/esp32.md#platforms) and [Pico](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/devices/pico.md#platforms) development boards, many of which have displays, that should also work.

**Comment by rei-vilo** [comment]
Thank you for the reference. I have noticed the Moddable Two. Are there distributors in Europe?

**Reply by phoddie** [reply]
Sorry, no distributor in Europe yet. The shipping costs are unfortunate. That said, we do regularly receive orders from that part of the world, so we should look into it again.

**Comment by rei-vilo** [comment]
Now that I managed to get the Moddable Two board working safely and nicely (plus a checklist for possible updates of the SDKs), I am back to this thread. 

The key point is to leverage the dashboard from Node-RED, as the editor provides a vast assortment of elements and an easy no code/low code implementation.

**Reply by ralphwetzel** [reply]
@rei-vilo 
To update you on the progress: The plugin meanwhile supports building flows with ui_nodes. We welcome you feedback on your test of this feature!

**Reply by rei-vilo** [reply]
Great!

**Reply by rei-vilo** [reply]
See #55 **Testing Node-RED dashboard on Moddable Two**.

**Comment by tve** [comment]
Q: does this use the std socket.io library? Or how does the MCU communicate with the browser?

**Reply by phoddie** [reply]
All I/O in Node-RED MCU Edition is done using ECMA-419 I/O. This includes HTTP, UDP, WebSocket, and MQTT. In the readme, many of the modules note the specific ECMA-419 they use ([UDP](https://github.com/phoddie/node-red-mcu/blob/main/readme.md#udp-in), [WebSocket In](https://github.com/phoddie/node-red-mcu/blob/main/readme.md#websocket-client), [HTTP Request](https://github.com/phoddie/node-red-mcu/blob/main/readme.md#http-request-1), etc).

(This question is out of place in this topic. If there are follow-up questions, please open a new discussion thread to help keep things organized. Thank you!)

**Reply by tve** [reply]
I had not realized that "Node-RED dashboard runs in MCU" meant that you display equivalent widgets on an attached display. I had thought it meant it could display to a regular Node-TED dashboard running in a browser that talks to the MCU. Hence the reference to socket.io...

**Reply by phoddie** [reply]
Thanks. I understand now why you posted this question on this thread. If you haven't already, do check out the [video](https://youtu.be/Inf37X0_xxo?t=8381) @rei-vilo linked to above. In the presentation, I show several examples of the dashboard running on a Moddable Two board powered by an ESP32. It works quite well.

**Reply by tve** [reply]
Yup, that's how I noticed my misconception... I'm now trying to install all this stuff to try it out myself (I mean node-red-mcu-plugin, not the dashboard part)...

**Reply by phoddie** [reply]
Cool. Good luck. @ralphwetzel has done some really nice work on the MCU plugin. If you run into issues getting set-up, Ralph has been really good about providing assistance. Just open an issue in his repo.
